src/kagglehub/gcs_upload.py

```diff
@@ -81,23 +82,32 @@ def _upload_blob(file_path: str, model_type: str):
 
 
 def upload_files(folder: str, model_type: str, quiet: bool = False):  # noqa: FBT002, FBT001
-    """upload files in a folder. Throws an exception if more than 50 files are found.
+    """upload files in a folder. Zips the files if there are more than 50.
     Parameters
     ==========
     folder: the folder to upload from
     quiet: suppress verbose output (default is False)
     model_type: Type of the model that is being uploaded.
     """
-    # TODO(b/312511716): Handle case where more than 50 files are to be uploaded by first zipping the content
 
     # Count the total number of files
     file_count = 0
     for _, _, files in os.walk(folder):
         file_count += len(files)
 
     if file_count > MAX_FILES_TO_UPLOAD:
-        max_files_to_upload_exception = "Cannot upload more than 50 files. Consider zipping the files first."
-        raise ValueError(max_files_to_upload_exception)
+        if not quiet:
+            logger.info(f"More than {MAX_FILES_TO_UPLOAD} files detected, creating a zip archive...")
+
+        zip_path = os.path.join(folder, "archive.zip")
```

Comment:
We should create the archive in a temporary folder for two reasons:

1. We don't want to mutate the directory specified by the user. This side-effect would be unexpected by the caller and could break their workflow.
2. If the user has a file named `archive.zip` in the folder they are trying to upload, then this would break.

You can use `from tempfile import TemporaryDirectory` for this.

Make sure to use the `with` statement to ensure the zip file is deleted after the upload completes:

```
with TemporaryDirectory() as d:
  // Create the zip
  // Upload the zip file
```